tox.ini

```diff
@@ -274,3 +274,34 @@ commands =
         --acme-server={env:ACME_SERVER:pebble}
 passenv = DOCKER_*
 setenv = {[testenv:py27-oldest]setenv}
+
+[testenv:travis-test-farm-tests-base]
+changedir = tests/letstest
+commands =
+    openssl aes-256-cbc -K {env:encrypted_9a387195a62e_key} -iv {env:encrypted_9a387195a62e_iv} -in travis-test-farm.pem.enc -out travis-test-farm.pem -d
+deps = -rtests/letstest/requirements.txt
+passenv = AWS_*
+setenv = AWS_DEFAULT_REGION=us-east-1
+whitelist_externals = openssl
+
+[testenv:travis-test-farm-tests-part1]
+changedir = {[testenv:travis-test-farm-tests-base]changedir}
+commands =
+    {[testenv:travis-test-farm-tests-base]commands}
+    python multitester.py apache2_targets.yaml travis-test-farm.pem none scripts/test_apache2.sh --fast
+    python multitester.py targets.yaml travis-test-farm.pem none scripts/test_leauto_upgrades.sh --fast
+deps = {[testenv:travis-test-farm-tests-base]deps}
+passenv = {[testenv:travis-test-farm-tests-base]passenv}
+setenv = {[testenv:travis-test-farm-tests-base]setenv}
+whitelist_externals = {[testenv:travis-test-farm-tests-base]whitelist_externals}
+
+[testenv:travis-test-farm-tests-part2]
```

Comment:
The part2 execution time is still extremely high, 48 minutes in the last run. We are not very far from the hard limit of 1 hour, when Travis just kill the entire job, and so potentially leaving here I suppose some AWS instances uncleaned.

Can we either:
1) Put some of the work here to part 1
2) Split part 2 again to have globally three parts ?

I think, for the global run efficiency and to avoid potential hard kill as I said just before, that all theses runs should have an execution time below 30 minutes.

Reply:
Unfortunately the times of these tests in general vary wildly. For instance, here's another run where part 2 is fast and part 1 is dangerously slow: https://travis-ci.com/certbot/certbot/builds/112169625

I haven't seen the builds time out with this configuration yet though. I've ran them a half a dozen times or so.

What I propose we do is do some more testing (which can happen either before or after this is merged) and if we (regularly?) timeout split the 4 tests into 4 individual jobs and deal with the consequences of doing that which I described above.

Reply:
What you describe about the big changes observed in execution time makes me think that in fact, the two parts are in a race condition to get AWS instances. I would bet that in fact one of the part is blocking until it gets the AWS instance because the other one is already holding all the nodes that can be run in parallel.

Could you move one part at the beginning of the Travis job list and the other one at the end? This way we will would be sure that the two parts are not running in parallel.

Reply:
Hm. If we think that's the problem, I'd be tempted to put them all in one Travis build. That build may take ~30 minutes, but it should minimize that problem.

I'm not entirely confident that's what's going on though. Here's a build where there were 4 tests running in parallel and they all were fast: https://travis-ci.com/certbot/certbot/builds/112154559

Reply:
Do you want to avoid having 4 jobs in parallel? Because it seems to be very efficient and keep, when every thing is good, the execution time very low. If in bad situations it takes 4 or 5 times more, we are still good.

Reply:
I did. My explanation was:

> The reason I put 2 scripts in the same build is there is a limit on the number of concurrent EC2 instances you can have. Our limit is 87 so to avoid problems due to two full test suites running at once and/or someone trying to do manual testing, I put them in two builds.

Looking at this problem more closely, this seems unlikely to be a problem though (as long as we do a good job making sure instances are properly cleaned up in the case of an error while running these tests such as them timing out). It seems unlikely that two nightly builds run at once fast enough to allow the test farm tests in different Travis builds to overlap. When running the client tests locally, running the tests in parallel on the same machine significantly slows down tests so Erica and I usually only run at most two tests at once.

Looking at those numbers, currently the 4 tests being run in parallel is 42 test instances plus 1 boulder instance which is shared. We have issues in the current milestone to potentially add 1 or 2 more instances to each run which could increase this to 50 client instances with the non-Apache test farm tests using 13 client instances. So even if we were running two more of these tests locally, that's only 50 + 13 * 2 + 1 = 77 instances so we should be fine.